File: java/src/us/palpant/science/kmc/plugins/NObjects.java
package us.palpant.science.kmc.plugins;

import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import org.apache.log4j.Logger;

import us.palpant.Ark;
import us.palpant.science.kmc.State;
import us.palpant.science.kmc.geometry.Lattice;

/**
 * Calculate the number of objects of a certain state on the Lattice
 * 
 * @author timpalpant
 *
 */
public class NObjects extends Plugin {
  
  private static final Logger log = Logger.getLogger(NObjects.class);
  
  private final State state;
  private final Path outputFile;
  private final PrintWriter writer;
  
  public NObjects(Lattice lattice, Ark config) throws IOException {
    super(lattice);
    state = State.forName((String)config.get("state"));
    outputFile = Paths.get((String)config.get("name"));
    log.debug("Opening nobjects output: "+outputFile);
    writer = new PrintWriter(Files.newBufferedWriter(outputFile, Charset.defaultCharset()));
  }
  
  @Override
  public void process(double time) {
    writer.println(time+"\t"+lattice.count(state));
  }

  @Override
  public void close() throws IOException {
    log.debug("Closing output file: "+outputFile);
    writer.close();
  }

}

File: cpp/kmc/nobjects.cpp
//
//  nobjects.cpp
//  kmc
//
//  Created by Timothy Palpant on 3/31/13.
//  Copyright (c) 2013 Timothy Palpant. All rights reserved.
//

#include "nobjects.h"

namespace kmc {
  namespace plugin {
    void NObjects::configure(const boost::property_tree::ptree& pt) {
      p_ = pt.get<boost::filesystem::path>("output");
      std::string particle = pt.get<std::string>("particle");
      state_ = lattice::State::for_name(particle);
    }
    
    void NObjects::boot(kmc::lattice::Lattice* lattice) {
      lattice_ = lattice;
      of_.open(p_.c_str());
      of_ << "# " << state_->name() << std::endl;
    }

    void NObjects::process(double time) {
      std::size_t count = 0;
      for (const lattice::State* state : lattice_->states()) {
        count += (state == state_);
      }
      of_ << time << '\t' << count << std::endl;
    }
    
    void NObjects::close() {
      of_.close();
    }
  }
}
